msal/src/main/java/com/microsoft/identity/client/tenantprofile/AccountAdapter.java

```diff
@@ -0,0 +1,261 @@
+        @Override
+        public List<ICacheRecord> filter(@NonNull final List<ICacheRecord> records) {
+            final List<ICacheRecord> result = new ArrayList<>();
+
+            // An account is orphaned if:
+            // 1. There is no home account which exists which has the same home_account_id
+
+            // First, get the home accounts....
+            final List<ICacheRecord> homeRecords =
+                    filterCacheRecords(
+                            records,
+                            new HomeAccountFilter(false)
+                    );
+
+            // Then, get the guest accounts...
+            final List<ICacheRecord> guestRecords =
+                    filterCacheRecords(
+                            records,
+                            new HomeAccountFilter(true)
+                    );
+
+            // Iterate over the guest accounts and find those which have no associated home account
+            for (final ICacheRecord guestRecord : guestRecords) {
+                if (hasNoCorrespondingHomeAccount(guestRecord, homeRecords)) {
+                    result.add(guestRecord);
+                }
+            }
+
+            return result;
+        }
+    };
+
+    /**
+     * For a supplied List of ICacheRecords, create each root IAccount based on the home
+     * account and then add child-nodes based on any authorized tenants.
+     *
+     * @param allCacheRecords
+     * @return
+     */
+    @NonNull
+    public static List<IAccount> adapt(@NonNull final List<ICacheRecord> allCacheRecords) {
+        // First, get all of the ICacheRecords for home accounts...
+        final List<ICacheRecord> homeCacheRecords = filterCacheRecords(
+                allCacheRecords,
+                new HomeAccountFilter(false)
+        );
+
+        // Then, get all of the guest accounts...
+        final List<ICacheRecord> guestCacheRecords = filterCacheRecords(
+                allCacheRecords,
+                new HomeAccountFilter(true)
+        );
+
+        final List<ICacheRecord> orphanCacheRecords = filterCacheRecords(
+                allCacheRecords,
+                orphanAccountFilter
+        );
+
+        // Remove the orphan records from the guest records...
+        guestCacheRecords.removeAll(orphanCacheRecords);
+
+        final List<IAccount> rootAccounts = createRootAccounts(homeCacheRecords);
+        appendChildren(rootAccounts, guestCacheRecords);
+        rootAccounts.addAll(createOrphanRoots(orphanCacheRecords));
+
+        return rootAccounts;
+    }
+
+    private static List<IAccount> createOrphanRoots(
+            @NonNull final List<ICacheRecord> orphanCacheRecords) {
+        // First, bucket the records by homeAccountId to create affinities
+        final Map<String, List<ICacheRecord>> bucketedRecords = new HashMap<>();
+
+        for (final ICacheRecord cacheRecord : orphanCacheRecords) {
+            final String cacheRecordHomeAccountId = cacheRecord.getAccount().getHomeAccountId();
```

Comment:
This should be the `oid`

Reply:
Keeping this bit of the code as-is - resolved in 9a753d8314a236f4d58e40b81a9a8b686a5910c0